### PR TITLE
fix(cli): patch glibc dependencies

### LINKED ISSUE #13516: deno upgrade to 1.18.1 fails on Ubuntu 16.04 with "GLIBC_2.27 not found"
This happens on Ubuntu Linux 16.04, 4.4.0-218-generic #251-Ubuntu SMP Thu Dec 2 09:48:54 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux:

```
$ RUST_BACKTRACE=1 deno upgrade
Looking up latest version
Found latest version 1.18.1
Checking https://github.com/denoland/deno/releases/download/v1.18.1/deno-x86_64-unknown-linux-gnu.zip
31.4 MiB / 31.4 MiB (100.0%)
Deno is upgrading to version 1.18.1
Archive:  /tmp/.tmpjvCQBk/deno.zip
  inflating: deno                    
/tmp/.tmpjvCQBk/deno: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.27' not found (required by /tmp/.tmpjvCQBk/deno)

============================================================
Deno has panicked. This is a bug in Deno. Please report this
at https://github.com/denoland/deno/issues/new.
If you can reliably reproduce this panic, include the
reproduction steps and re-run with the RUST_BACKTRACE=1 env
var set and include the backtrace in your report.

Platform: linux x86_64
Version: 1.18.0
Args: ["deno", "upgrade"]

thread 'main' panicked at 'assertion failed: output.status.success()', cli/tools/upgrade.rs:306:3
stack backtrace:
   0: rust_begin_unwind
             at /rustc/02072b482a8b5357f7fb5e5637444ae30e423c40/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /rustc/02072b482a8b5357f7fb5e5637444ae30e423c40/library/core/src/panicking.rs:107:14
   2: core::panicking::panic
             at /rustc/02072b482a8b5357f7fb5e5637444ae30e423c40/library/core/src/panicking.rs:48:5
   3: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   4: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   5: deno::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Only the older  GLIBC 2.23 exits:

```
$ ldd --version
ldd (Ubuntu GLIBC 2.23-0ubuntu11.3) 2.23
```

deno starting with 1.18.1 will not run on a plain Ubuntu 16.04 anymore?

### PR DESCRIPTION
**Pull request by littledivy**
Temporary solution until https://github.com/lifthrasiir/hexf/pull/21 is merged and updated across dependencies.

Before:
```
nm -S (which deno) | grep exp2
00000001001a7c88 0000000000000000 t __ZN9deno_lint5rules17no_invalid_regexp22NoInvalidRegexpVisitor11check_regex17h138808595993f4e7E
00000001001aae08 0000000000000000 t __ZN9deno_lint5rules17no_invalid_regexp22NoInvalidRegexpVisitor23handle_call_or_new_expr17ha1cbb9245d2d5e30E
00000001001a7fe8 0000000000000000 t __ZN9deno_lint5rules17no_invalid_regexp22NoInvalidRegexpVisitor25check_for_invalid_pattern17h96851b015489f16aE
00000001001aaeec 0000000000000000 t __ZN9deno_lint5rules17no_invalid_regexp29check_expr_for_string_literal17h6f1f418555ce44c8E
                                  U _exp2f
```

After: 
```
nm -S target/release/deno | grep exp2
00000001028ef988 0000000000000000 s __ZN4libm4math4exp23TBL17h0075c2f89728ff9aE
00000001003dba28 0000000000000000 t __ZN9deno_lint5rules17no_invalid_regexp22NoInvalidRegexpVisitor11check_regex17h7d939a0c88dba0f9E
00000001003de8c8 0000000000000000 t __ZN9deno_lint5rules17no_invalid_regexp22NoInvalidRegexpVisitor23handle_call_or_new_expr17h146e6fd7dbfde289E
00000001003dbd80 0000000000000000 t __ZN9deno_lint5rules17no_invalid_regexp22NoInvalidRegexpVisitor25check_for_invalid_pattern17h1798b638cb488088E
00000001003de9ac 0000000000000000 t __ZN9deno_lint5rules17no_invalid_regexp29check_expr_for_string_literal17h20ec3079f65956d9E
```

Closes #13516

### CODE REVIEW

**Review by piscisaureus — APPROVED** [review]
Discussed offline. LGTM, pending verification that the [binary](https://drive.google.com/file/d/1H8UR5wEPXk7KsESDzjS9t-FLEroNnbEH/view) works on Amazon Linux 2.

File: .cargo/config [hunk 1]
```diff
@@ -1,2 +1,5 @@
 [target.x86_64-pc-windows-msvc]
 rustflags = ["-C", "target-feature=+crt-static"]
+
+[patch.crates-io]
```
Comment (piscisaureus) [inline .cargo/config:4]: Please add a comment to explain why this is needed.

**Comment by hayd** [comment]
@piscisaureus the binary you shared seems to give the same error, at least using Docker:
```Dockerfile
# Dockerfile
FROM amazonlinux:2.0.20200722.0
ADD deno /bin/deno
ENTRYPOINT deno
```


```sh
# Copy binary to directory (ensure chmod -x).
❱ docker build --platform=linux/amd64 -t glib2fix .
❱ docker run --platform=linux/amd64 --init -i -t glib2fix
/usr/bin/deno: /lib64/libc.so.6: version `GLIBC_2.27' not found (required by /usr/bin/deno)
```

**Comment by piscisaureus** [comment]
> @piscisaureus the binary you shared seems to give the same error, at least using Docker:

Ugh. It seems that 1.19.2 also introduced a dependency on the "copy_file_range" function.

**Comment by lucacasonato** [comment]
Will the git dependency prevent `cargo publish`?

**Comment by piscisaureus** [comment]
> Will the git dependency prevent cargo publish?

No, .cargo/config is ignored when publishing.

**Comment by littledivy** [comment]
```
[ec2-user@ip-172-31-84-146 ~]$ ./deno
./deno: /lib64/libc.so.6: version `GLIBC_2.27' not found (required by ./deno)
[ec2-user@ip-172-31-84-146 ~]$ objdump -T deno | grep -E 'GLIBC_2.[23][0-9]'
0000000000000000  w   DF *UND*0000000000000000  GLIBC_2.27  copy_file_range
0000000000000000  w   DF *UND*0000000000000000  GLIBC_2.25  getrandom
```

`copy_file_range` :(

**Comment by littledivy** [comment]
Apparently, Rust's libstd is built against a GLIBC with `copy_file_range`. We must use the `copy_file_range` syscall on older GLIBC, it's supported on Linux 4.5 and above. 

I've patched the symbol in https://github.com/denoland/deno/pull/13832/commits/2e2dcc03686a21545f058cf6399f8542c4084330

**Comment by littledivy** [comment]
Works on Amazon Linux 2

**Review by bartlomieju — APPROVED** [review]

**Comment by ry** [comment]
Maybe we should have a test for Amazonlinux so we don’t regress

**Comment by piscisaureus** [comment]
This PR hasn't landed it because I'm trying a less hacky approach that uses an ubuntu 16.04 sysroot.

When deno runs on that I believe it'll also run on Amazon linux.

**Comment by bartlomieju** [comment]
Should this PR be closed in favor of https://github.com/denoland/deno/pull/13846 ?

**Pull request closed without merging**